blueocean-dashboard/src/main/js/creation/bitbucket/cloud/BbCloudFlowManager.js

```diff
@@ -0,0 +1,414 @@
+import BbUnknownErrorStep from '../steps/BbUnknownErrorStep';
+import BbRenameStep from '../steps/BbRenameStep';
+
+const LOGGER = logging.logger('io.jenkins.blueocean.bitbucket-cloud-pipeline');
+const MIN_DELAY = 500;
+const FIRST_PAGE = 1;
+const PAGE_SIZE = 100;
+const SSE_TIMEOUT_DELAY = 1000 * 60;
+const translate = i18nTranslator('blueocean-dashboard');
+
+export default class BbCloudFlowManager extends FlowManager {
+
+    credentialManager = null;
+    queuedIndexAllocations = 0;
+
+    get credentialId() {
+        return this.credentialManager && this.credentialManager.credentialId;
+    }
+
+    @observable
+    organizations = [];
+
+    @observable
+    repositories = [];
+
+    @observable
+    repositoriesLoading = false;
+
+    @computed get selectableRepositories() {
+        if (!this.repositories) {
+            return [];
+        }
+        return this.repositories;
+    }
+
+    @observable
+    selectedOrganization = null;
+
+    @observable
+    selectedRepository = null;
+
+    @computed get stepsDisabled() {
+        return this.stateId === STATE.STEP_COMPLETE_SAVING_ERROR ||
+            this.stateId === STATE.STEP_COMPLETE_EVENT_ERROR ||
+            this.stateId === STATE.STEP_COMPLETE_EVENT_TIMEOUT ||
+            this.stateId === STATE.STEP_COMPLETE_MISSING_JENKINSFILE ||
+            this.stateId === STATE.STEP_COMPLETE_SUCCESS;
+    }
+
+    pipeline = null;
+
+    _repositoryCache = {};
+
+    _creationApi = null;
+
+    _sseSubscribeId = null;
+
+    _sseTimeoutId = null;
+
+    constructor(creationApi, credentialsApi) {
+        super();
+        this._creationApi = creationApi;
+        this.credentialManager = new BbCredentialManager(credentialsApi);
+    }
+
+    translate(key, opts) {
+        return translate(key, opts);
+    }
+
+    getApiUrl() {
+        return 'https://bitbucket.org';
+    }
+
+    getStates() {
+        return STATE.values();
+    }
+
+    getState() {
+        return STATE;
+    }
+
+    getInitialStep() {
+        return {
+            stateId: STATE.PENDING_LOADING_CREDS,
+            stepElement: <BbLoadingStep />,
+        };
+    }
+
+    onInitialized() {
+        this.findExistingCredential();
+        this.setPlaceholders(translate('creation.core.status.completed'));
+    }
+
+    destroy() {
+        this._cleanupListeners();
+    }
+
+    findExistingCredential() {
+        return this.credentialManager.findExistingCredential(this.getApiUrl())
+            .then(waitAtLeast(MIN_DELAY))
+            .then(success => this._findExistingCredentialComplete(success));
+    }
+
+    _findExistingCredentialComplete(success) {
+        if (success) {
+            this.changeState(STATE.PENDING_LOADING_ORGANIZATIONS);
+            this.listOrganizations();
+        } else {
+            this.renderStep({
+                stateId: STATE.STEP_CREDENTIAL,
+                stepElement: <BbCredentialsStep />,
+            });
+        }
```

Comment:
Right, but in the `BbServerFlowManager` class, this method will still fire after the credentials are fetched, and if no credentials are available then a `renderStep` method call without an `afterStateId` will remove any steps already registered. If you fire up BB Server, make sure all credentials are removed, then pick a server I think you'll see what I'm talking about, but I could be wrong. I don't have a BB server handy to test against.